### ISSUE TITLE
Convert `TestDelegate` and `ActualValueDelegate` to `Action` and `Func<T>`

### ISSUE THREAD

**Issue by stevenaw** [nunit/nunit#4824]
### Discussed in https://github.com/nunit/nunit/discussions/4823

<div type='discussions-op-text'>

<sup>Originally posted by **Dreamescaper** September 13, 2024</sup>
Is there any reason not to use `Action` and `Func<T>` instead? Since `Action` and `Func<T>` are natural types for delegates in C#, using other delegates causes some confusion, e.g. here:
```csharp
        var getActual = () => "/* get actual value */";
        Assert.That(getActual, Is.EqualTo("expectedValue").After(10000, 500));
```

Since I'm passing a delegate to Assert method, I'd expect a delegate overload to be invoked. Instead, non-delegate `TActual` overload is used, and assertion fails - because `getActual` is `Func<string>`, not `ActualValueDelegate<TActual>`.

Therefore, my question is. Would it make sense to to replace `TestDelegate` and `ActualValueDelegate` with `Action` and `Func<T>` in the next major version?</div>

**Comment by OsirisTerje** [comment]
@stevenaw @manfred-brands 

I am concerned if we remove the TestDelegate, as it is heavily used in the OS community (and probably in a lot of closed source too). That means if we simply replace it , it may inconvenience a lot of people.  (We got a lot of flak when we just changed the AreEqual classname ...)

Using this search term `/\bTestDelegate\b/ ( "using NUnit.Framework" OR "NUnit.Framework" ) language:C#` it finds more than 4.9K hits on github.  

In many cases (like inline lambdas) a replacement will work, but when we have variables that are TestDelegates, it will not.

If we keep the TestDelegate, we will cause ambiguity in cases where Assert.That will have both TestDelegate and Func/Action, again lambdas will not compile here.

Can we find a way to smooth the transfer ?  
Do we have to use That as the keyword - it is pretty overloaded already ?   
Could we for the ACtion/Func overloads introduce a (just off the top of my head) Assert.Call(  or Assert.ThatCall( ?

**Comment by manfred-brands** [comment]
@OsirisTerje I hadn't anticipated that users actually would use these delegate types in their own code.

We can easily keep the definitions IFF the code would still work, but it doesn't.

```csharp
string GetExceptionMessage() => $"Not Equal to {4}";
ActualValueDelegate<int> actual = () => 2 + 2;
Assert.That(actual, Is.EqualTo(4), GetExceptionMessage);
```

Although the above compiles, it doesn't call the right overload.
`ActualValueDelegate<int>` doesn't match `Func<int>` nor can it be cast to it.

So we would need to duplicate the overloads having both `Func<TActualValue>` and `ActualValueDelegate<TActual>`.
Which doesn't work either as these are ambiguous for inline lambdas.

We cannot make an alias `ActionDelegate<T>` == `Func<T>`

So we cannot do this without breaking code. 

Introducing an `Assert.Call` is unnecessary.
The whole reason for this would be to standardize on pre-defined types instead of our own.

I suggest then to close this ticket as "cannot be done".

**Comment by OsirisTerje** [comment]
I tend to agree, but it is brutal.  Hard stops like this is not fun.
Could we find out how widespread the usage is ?   By e.g. using the analyzer and raise a warning or a hint ?  
As it is today we also have an annoying issue in the opposite direction, we don't accept any Func/Action variables, that gives a type mismatch compiler error, which is also a kind of "design by surprise".

I found these scenarios:

| Scenario                              | Works today                     | Works if replaced |
| ------------------------------------- | ------------------------------- | ----------------- |
| Inline lambda                         | ✅                              | ✅                 |
| `Action` / `Func<T>` (inline lambda)  | ✅ (via implicit conversion)     | ✅                 |
| `Action` / `Func<T>` (as variable)    | ❌ (type mismatch)              | ✅                 |
| Explicit `TestDelegate` usage         | ✅                              | ❌ (breaking)      |

Examples of explicit TestDelegate usages are:

- Explicit variable (most common)
- Method return types
- Method parameters / helpers
- Fields / properties
- Extension / library code (important) (3rd party helpers or frameworks)

Not sure how much there is of either of these

@jnm2 Can you do anything language wise ?

@stevenaw @CharliePoole @rprouse Comments ?

**Comment by stevenaw** [comment]
This is unfortunate and surprising to me. 

My initial thoughts here would be to ease conversion using implicit conversion operators, but I think (as I am without a computer right now) those can not be defined for delegate types nor as extension members. I have found this discussion where the latest update as of Oct 2025 was that it could be considered risky to implement as extension members https://github.com/dotnet/csharplang/discussions/515

**Comment by stevenaw** [comment]
An idea: Delegates can inherit from each other. Could our delegates have their definitions updated so that they inherit from the built-in types we want to move towards? If that were possible then it "might" also allow for implicit conversions to work via the inheritance chain.

**Comment by OsirisTerje** [comment]
@manfred-brands   `ActualValueDelegate` has only 291 hits, and most of them are either nunit or forked nunit repos.  
It seems the main culprit is the `TestDelegate`.

**Comment by OsirisTerje** [comment]
@stevenaw   C# does not allow delegates to inherit from other delegates, so this path would then be closed.


Ref https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/delegates/using-delegates  
>Delegate types are derived from the [Delegate](https://learn.microsoft.com/en-us/dotnet/api/system.delegate) class in .NET. Delegate types are [sealed](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/sealed), they can't be derived from, and it isn't possible to derive custom classes from [Delegate](https://learn.microsoft.com/en-us/dotnet/api/system.delegate).

**Comment by stevenaw** [comment]
Ah. Right you are. I had misread some code examples here prior to making that earlier comment

https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/covariance-contravariance/variance-in-delegates

Delegate inheritance would be a nice language feature that could help us here if only it were supported.

**Comment by CharliePoole** [comment]
Given the language limitations, what about just adding additional overloads to `Assert.That` -  and any other methods that use these Types?

**Comment by OsirisTerje** [comment]
It will be ambiguity crashes:   "If we keep the TestDelegate, we will cause ambiguity in cases where Assert.That will have both TestDelegate and Func/Action, again lambdas will not compile here."

**Comment by CharliePoole** [comment]
Ah, I missed that. Then I think getting rid of TestDelegate  as a breaking change for 5.0 makes sense.

**Comment by OsirisTerje** [comment]
TestDelegate has  4900 hits on github.  It is heavily used "out there".  Do we want to inconvenience so many?  Last time we did, we got a lot of flak.

**Comment by stevenaw** [comment]
I suspect our focus and discussion has (rightly) focused on the more-visible changes to the assertion API. I noticed there's also some references within constraint signatures. Those strike me as much lower risk to update as they'd likely only affect those who have written custom constraints. They'd also likely be _easier_ to update in a non-breaking way as it's unlikely they are invoked directly outside of NUnit and less-likely still to be done so using an inline lambda function. We could likely begin moving away from those even in a 4.x release by just adding a virtual (not abstract) `Func` or `Action` based overload on the constraint and marking the other `obsolete`.

**Comment by manfred-brands** [comment]
> It will be ambiguity crashes: "If we keep the TestDelegate, we will cause ambiguity in cases where Assert.That will have both TestDelegate and Func/Action, again lambdas will not compile here."

I just remembered something. This is why [OverloadResolutionPriorityAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.overloadresolutionpriorityattribute?view=net-10.0) was invented.

By duplicating the asserts using both `TestDelegate` and `Action` and then put an `[OverloadResolutionPriority(1)]` on the new method thereby prioritizing the new overload.

This seems to work.

Initially I had marked the `AsyncTestDelegate` overloads as priority -1, but that would mean even if I had an `AsyncTestDelegate`, it would call the `<TActual>` overload as it had a higher priority than the real one.

I also needed to mark `Func<Task<T>>` overloads as priority 2 as otherwise it would call the `Func<T>` overload instead.

**Comment by CharliePoole** [comment]
@manfred-brands That's actually pretty cool. I didn't know about the attribute. Nevertheless, I wonder what would be the benefit of retaining TestDelegate at all if you're talking about a major release where breaking changes are to be expected.

**Comment by OsirisTerje** [comment]
@manfred-brands You're gold !  ❤️ 

That attribute is a life-saver.  I was not at all aware of it, not a clue!   Awesome!  
Then we just proceed.

I think marking TestDelegate Obsolete is good, because we can then use that to point people on how to resolve it.   

It also means we can move them gradually over to the new functionality, and perhaps we can then even implment this in the 4.X series, and use that as a way forward, and in 5.X remove the TestDelegate - or we could let it be for a while longer and remove it in 6.X, as long as it doesn't disturb us.

**Comment by manfred-brands** [comment]
>  I wonder what would be the benefit of retaining TestDelegate at all if you're talking about a major release where breaking changes are to be expected.

@CharliePoole There is expected with warning and out of the blue.

We marking it obsolete in 5.X and remove it in 6.X.
We could mark it obsolete in 4.6 and remove it in 5.X except for the one incompatible changes as mentioned in the PR comments.

**Comment by stevenaw** [comment]
Nice thinking about `OverloadResolutionPriorityAttribute` @manfred-brands 
I notice that it was introduced in .NET 9. Would that also work for .NET Framework consumers?

**Comment by manfred-brands** [comment]
> I notice that it was introduced in .NET 9. Would that also work for .NET Framework consumers?

It is purely a compiler feature. As long as they use an SDK for .net 9 or later. But as we already use C#14 extension methods we require the .NET 10 SDK anyhow

**Comment by CharliePoole** [comment]
@manfred-brands OK, if this is the general view of the team about major releases then that makes sense.

One reservation...

IMO, such a conservative approach is more appropriate for a project with frequent major releases - i.e. every year or two or less. NUnit is (has been) a project that does major releases very infrequently. Version 4.0 was released 23 years after version 1.0. So maybe it's a good idea to discuss when we expect to see 5.0, 6.0, etc. in general terms.

I think it would be OK to change, but it should be talked about. Some projects do an annual major release or the equivalent of one if they are not using Semver. For a project that works that way and also stops doing maintenance fixes on the previous year's release, then I agree that a very conservative approach to breaking is needed.

FWIW in the 4.0 Engine, we started out with a view that many breaking changes should be gotten out of the way as quickly as possible. We've been very open about that but even so I'm sure some users will stick with 3.x for years after 4.0 is released - exactly as happened in moving from 2.0 to 4.0.

So bottom line... are we thinking of 5.0 this year, next year or when? Same question for 6.0...

UPDATE: Obviously I'm broadening the discussion way beyond this one issue. Maybe we should discuss elsewhere. :-)

**Comment by OsirisTerje** [comment]
We have started to try to start a release when we have more than 20 closed issues.  Normally that has been minor releases.  Also, that we should try to release approx 4 times a year.  Version 4.5 had 40 issues, 4.4 had 26, 4.3 had 29, all with approx 6 months in between.

I think we can go with the same with the major releases, except not that cadence, perhaps once a year, but it should also be above a certain level of issues resolved, or some major changes we find we need to push out. 

With todays pace in development, I think we should try to keep ut.  

And, it's fun!

**Comment by CharliePoole** [comment]
@OsirisTerje That's a reasonable decision to make but I think you have to address how to make a transition from one major release every 8 to 10 years to a more frequent cycle.

If you were to do 5.0 this year, that's three years since 4.0, so shorter than previous but not as short as annual. So maybe 5.0 needs slightly different criteria from 6.0, which would hypothetically come in 2027. Maybe we announce this new approach at the same time as you do the 5.0 release or even with the 4.6 release.

If we apply the same approach to the engine and console, 4.0 will be coming out this year (I hope), 11 years after 3.0. However, all the breaking changes have been announced widely and some (but not all) have been deprecated in 3.X. So for that project a different approach may be called for... maybe a final 3.X release with added deprecation of features missing in 4.0.

The kind of announcement I'm talking about might be something like...

"In addition to this new major release of the NUnit (framework | console and engine | adapter), we wanted to let you know that we plan to increase the pace of new major releases so as to introduce new features more rapidly. The next major release of the (framework | engine  | etc.), version X.X is anticipated in MONTH, YEAR."

It could be less specific if we don't know when it's expected, of course.

**Comment by OsirisTerje** [comment]
"Moved" the release cadence discussion here:  https://github.com/nunit/nunit/discussions/5223

**Comment by OsirisTerje** [comment]
@manfred-brands 

I searched around to find a public repo that used TestDelegates, and was not a nunit clone. I found the SharpKml - I have no knowledge of this project from earlier.  Just wanted something "out there" to test against.

I then cloned the https://github.com/samcragg/sharpkml (net8.0) which I know use some TestDelegates, and ran that with the latest code, and then got a lot of CS0121:

```txt
    C:\repos\nunit\tds\sharpkml\UnitTests\Base\KmlFactoryTest.cs(168,24): error CS0121: The call is ambiguous between the following methods or properties: 'NUnit.Framework.Assert.That(NUnit.Framework.TestDelegate, NUnit.Framework.Constraints.IResolveConstraint, NUnit.Framework.NUnitString, string, string)' and 'NUnit.Framework.Assert.That(System.Action, NUnit.Framework.Constraints.IResolveConstraint, NUnit.Framework.NUnitString, string, string)'
    C:\repos\nunit\tds\sharpkml\UnitTests\Dom\ElementTest.cs(40,24): error CS0121: The call is ambiguous between the following methods or properties: 'NUnit.Framework.Assert.That(NUnit.Framework.TestDelegate, NUnit.Framework.Constraints.IResolveConstraint, NUnit.Framework.NUnitString, string, string)' and 'NUnit.Framework.Assert.That(System.Action, NUnit.Framework.Constraints.IResolveConstraint, NUnit.Framework.NUnitString, string, string)'
    C:\repos\nunit\tds\sharpkml\UnitTests\Base\KmlFactoryTest.cs(131,24): error CS0121: The call is ambiguous between the following methods or properties: 'NUnit.Framework.Assert.That(NUnit.Framework.TestDelegate, NUnit.Framework.Constraints.IResolveConstraint, NUnit.Framework.NUnitString, string, string)' and 'NUnit.Framework.Assert.That(System.Action, NUnit.Framework.Constraints.IResolveConstraint, NUnit.Framework.NUnitString, string, string)'
```

This is because the [OverloadResolutionPriorityAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.overloadresolutionpriorityattribute?view=net-10.0) attribute came with net 9.0.  So with net 8.0 it doesn't work.  

This is a compiler feature, so if I change the test project to target net 10, it works as it should, and with the obsolete warnings too. (And since it is a compiler feature, it works with the net8 lib).  I has exactly the same number of tests as before I upgraded to the new 4.6, so it actually works!  That is cool!

```txt
  UnitTests net10.0 succeeded with 2 warning(s) (1.2s) → UnitTests\bin\Debug\net10.0\UnitTests.dll
    C:\repos\nunit\tds\sharpkml\UnitTests\Engine\KmzFileTest.cs(65,30): warning CS0618: 'TestDelegate' is obsolete: 'Use Action instead of TestDelegate'
    C:\repos\nunit\tds\sharpkml\UnitTests\Engine\KmzFileTest.cs(65,17): warning CS0618: 'Assert.That(TestDelegate, IResolveConstraint, NUnitString, string, string)' is obsolete: 'Use overload with Action instead of TestDelegate'
```

Then comes the issue - what about when they target net 8 or net 6 ?   As we don't (yet) have a net 10 library included, we can't use conditional compiling to not include the new features for net 8 and net 6.  

If we include a library package for net 10 (it could be net 9 too, but that is not LTS - and its outdated this autumn. ), we could use conditional compiling to exclude the new functionality for net 8 and net 6.  And with v5 we will skip net 6 anyway, but we will still have net 8 for quite some time.  So I think we need to do this in 4.6.

And, it will not work with net462 either, so we need to use conditional compiling here. That attribute will not come to net framework.

About net 9, I am fine with just stating that we don't support delegates in net 9.  Or we could just ship a net 9 library too, it is not that much of a change anyway. 

I can't see any other way to handle this.

PS:  In V5 we can skip TestDelegates and its friend altogether. 
PS2: For all frameworks we should keep the Obsolete, but also add it require net 10 in version 4.6+.  Or say that it will be removed in v5. 
PS3: Net 8 is out of support Nov 2026, so as we have said earlier, we can keep it for 6 more months and then skip it.

**Comment by manfred-brands** [comment]
@OsirisTerje You should only have to set global.json to SDK 9+ to get the benfit, there should be no need to target .NET 9+.

I can try that repository tomorrow.

**Comment by OsirisTerje** [comment]
@manfred-brands I tried that, but that doesn't work.  It seems to be more than **just** a compiler feature, it is somehow part of the BCL too. 

From ChatGPT:

> Even though the compiler comes from the SDK, Roslyn does feature gating based on the target framework.

>For this specific feature:

>It is not purely a compiler feature
>It is treated as a framework feature introduced in .NET 9

**Comment by manfred-brands** [comment]
@OsirisTerje It depends on which AI you ask: From Gemini:

> The [OverloadResolutionPriorityAttribute](https://www.google.com/search?q=OverloadResolutionPriorityAttribute&ved=2ahUKEwix5tPqxv2TAxVcr1YBHakBL9QQgK4QegQIARAB) (introduced in C# 13) does not natively work with the .NET 8 SDK and C# 12 compiler, as the feature requires the C# 13 compiler. 

> However, you can use it in a .NET 8 project if you fulfill specific requirements to enable C# 13 features. 
How to Use it in .NET 8
To make [OverloadResolutionPriority] work in a .NET 8 project, you must:
Use the Visual Studio 17.12+ or .NET 9 SDK (which includes the C# 13 compiler).
Set your project to use C# 13, even while targeting .NET 8:

The trick is to set the `LangVersion` to 13+.
If I do that I get 2 warnings:

```text
  UnitTests net8.0 succeeded with 2 warning(s) (0.4s) → bin\Debug\net8.0\UnitTests.dll
    D:\Development\3rd\samcragg\sharpkml\UnitTests\Engine\KmzFileTest.cs(65,30): warning CS0618: 'TestDelegate' is obsolete: 'Use Action instead of TestDelegate'
    D:\Development\3rd\samcragg\sharpkml\UnitTests\Engine\KmzFileTest.cs(65,17): warning CS0618: 'Assert.That(TestDelegate, IResolveConstraint, NUnitString, string, string)' is obsolete: 'Use overload with Action instead of TestDelegate'
```

With unit tests succeeding on .net 8:
```text
NUnit Adapter 4.5.0.0: Test execution started
Running all tests in D:\Development\3rd\samcragg\sharpkml\UnitTests\bin\Debug\net8.0\UnitTests.dll
   NUnit3TestExecutor discovered 227 of 227 NUnit test cases using Current Discovery mode, Non-Explicit run
NUnit Adapter 4.5.0.0: Test execution complete
  UnitTests test net8.0 succeeded (1.1s)

Test summary: total: 227, failed: 0, succeeded: 227, skipped: 0, duration: 1.1s
```

So it seems to work if you use and SDK >= 9.0 and set the LangVersion >= 13

**Comment by OsirisTerje** [comment]
Very good - that is an "easy" way out, but it require the user to do something.  We can't enforce it directly.  
So, we then need to make the documentation of how to do this very visible for the user.

Or, can we add something to the analyzer that surface the information ?

**Comment by manfred-brands** [comment]
I can get the LanguageVersion in the Analyzer.
Do you want me to see if I can make a rule that if NUnit 4.6 (or higher) is used with a LangVersion < 13 that it reports an error explaining the possible CS0618 warning?
I would also need to determine the SDK because if LangVersion==Default the actual version depends on the SDK.

**Comment by OsirisTerje** [comment]
Yes, I think that can be helpful.  
Have to also cover LangVersion latest ofc. 

Should it be a general rule, or just if TestDelegate or ActualValueDelegate is used ?   The latter would be preferable.

**Comment by manfred-brands** [comment]
> Should it be a general rule, or just if TestDelegate or ActualValueDelegate is used ? The latter would be preferable.

If that is used, they get an Obsolete warning.

Ideally it should be when the CS0121 below is raised, however I don't know if the analyzer is called when the compiler generates errors. I'll find out.

> error CS0121: The call is ambiguous between the following methods or properties: 'NUnit.Framework.Assert.That(NUnit.Framework.TestDelegate, NUnit.Framework.Constraints.IResolveConstraint, NUnit.Framework.NUnitString, string, string)' and 'NUnit.Framework.Assert.That(System.Action, NUnit.Framework.Constraints.IResolveConstraint, NUnit.Framework.NUnitString, string, string)'

**Comment by OsirisTerje** [comment]
> [!IMPORTANT]
> This issue is a partially breaking change.  
> 
> If you use TestDelegate and Action **explicitly** , you must either:
> * Alternative 1:  Set your target framework to  .Net 9.0 or higher
> * Alternative 2:  You must set your language version to 13 or higher, and use the Visual Studio 2022 version 17.12 or higher, or use .NET 9 SDK (which includes the C# 13 compiler), while still targeting .net8 or lower.
> 
> | Scenario                              | NUnit pre 4.6 | NUnit 4.6+ (.net < 9.0)  | NUnit 4.6+ (.net >= 9.0) | NUnit 4.6+ (.net<9.0 and  langversion >=13) |
> | ------------------------------------- | ------------------------------- | ----------------- | ------ | ----- |
> | Inline lambda                         | ✅                              | ❌ (breaking)                   | ✅                 | ✅                 |
> | `Action` / `Func<T>` (inline lambda)  | ✅ (via implicit conversion)     | ❌ (breaking)                  | ✅                 |✅                 |
> | `Action` / `Func<T>` (as variable)    | ❌ (type mismatch)              | ✅                 | ✅                 |✅                 |
> | Explicit `TestDelegate` usage         | ✅                              | ❌ (breaking)      | ✅                 |✅                 |
> | Evaluates functions | ❗  (success, but no eval) |  ✅ | ✅|✅                 |
> ---
> Examples of explicit TestDelegate usages are:
> 
> - Explicit variable (most common)
> - Method return types
> - Method parameters / helpers
> - Fields / properties
> - Extension / library code (important) (3rd party helpers or frameworks)
> 
> Code for reproduction of these cases can be found at https://github.com/nunit/nunit.issues/tree/main/Issue4824

**Comment by manfred-brands** [comment]
@OsirisTerje

If not setting the LangVersion do we not get ambiguous errors for any use of an anonymous lambda?

**Comment by OsirisTerje** [comment]
@manfred-brands Not sure, but I tested earlier, and got it to work.  But might be we had more changes or I missed something.  Will test again tomorrow.  Preparing release notes now, and will include the IMPORTANT section above there. The GetChanges tool picks it up automatically, so we should try to get it accurate.

**Comment by OsirisTerje** [comment]
@manfred-brands

You're right about the :

> If not setting the LangVersion do we not get ambiguous errors for any use of an anonymous lambda?

So I have tested all variations (see  https://github.com/nunit/nunit.issues/tree/main/Issue4824)  and also added the changed functionality to the table above.  Also added 4.13 of the analyzer.

**Comment by Smaug123** [comment]
For reference (since Dependabot is producing so many backlinks that the automatic reference would otherwise get lost), this change was more breaking than I think you expected; see https://github.com/nunit/nunit/issues/5252 .

**Comment by OsirisTerje** [comment]
Seems many (or most) of these are analyzer updates.  Not sure why dependabot add these to just this issue.   

I have just checked a few, but it seems the updates works for the great majority of these.

**Comment by giesen-ascora** [comment]
I have hundreds of assertions that look something like this:
```
Assert.That(() => _sut.DoSomethingStupid(), Throws.Exception);
```
None of these compile anymore after the upgrade to 4.6.0. I get compiler errors like:
> The call is ambiguous between the following methods or properties: 'Assert.That<TActual>(ActualValueDelegate<TActual>, IResolveConstraint, NUnitString, string, string)' and 'Assert.That<TActual>(Func<TActual>, IResolveConstraint, NUnitString, string, string)'

According to what you wrote above, inline lambdas were still supposed to work, but they don't...  
How do I have to resolve this? Do I really have to add explicit casts to `Action` all over the place?  
BTW: This is in a solution that's still using .NET Framework 4.8 .

**Comment by manfred-brands** [comment]
@giesen-ascora If you use an SDK 9 or later you need to set LangVersion to 13.

This is described in the release notes under [breaking changes](https://docs.nunit.org/articles/nunit/release-notes/framework.html#the-following-issues-are-marked-as-breaking-changes)

**Comment by giesen-ascora** [comment]
> [@giesen-ascora](https://github.com/giesen-ascora) If you use an SDK 9 or later you need to set LangVersion to 13.
> 
> This is described in the release notes under [breaking changes](https://docs.nunit.org/articles/nunit/release-notes/framework.html#the-following-issues-are-marked-as-breaking-changes)

As I wrote, this is in .NET Framework 4.8 projects (C# 7.3), so I'm afraid that's not an option for me...

**Comment by manfred-brands** [comment]
> As I wrote, this is in .NET Framework 4.8 projects (C# 7.3), so I'm afraid that's not an option for me...

You do NOT have to change your TargetFramework.
It is possible to use most of the C# 14 features in .NET Framework.

**Comment by Otiel** [comment]
I'm in the same boat as @giesen-ascora : my solution targets .NET Framework 4.8, and I currently use C# 8.0.

I'm surprised that NUnit advises to set the language version to 13.0 or higher as Microsoft [clearly states](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-versioning) that using C#13 on .NET Framework 4.8 is unsupported:
  > Choosing a language version newer than the default can cause hard-to-diagnose compile-time and runtime errors.
  > ...
  > C# 13 is supported only on .NET 9 and newer versions. Using a C# language version newer than the version associated with your target TFM is unsupported.

I think it is not trivial to update the language version. For instance, even if my solution builds fine after upgrading to C# 13.0, I have some code inspection rules that do not pass anymore with around 13 000 errors. The number is not really meaningful, but I want to emphasize the fact that changing the language version is a breaking change that is not quickly / easily performed depending on the solution.

1. It feels like too big of a breaking change for a Minor upgrade.
2. Even after I fix my 13 000 errors, I'm not confident using a language version unsupported for my target framework as per Microsoft statement.

Is there any way to reconsider this change and go another route?

**Comment by giesen-ascora** [comment]
@Otiel : I have also put off upgrading to 4.6.x because of this. If I understood correctly, this is only an intermediary problem until the now deprecated `TestDelegate` and `ActualValueDelegate` have been properly removed from the code base, right?

**Comment by OsirisTerje** [comment]
@Otiel  @giesen-ascora  4.6.1 is a bugfix version, so that doesn't help with the issue you have.

Even C# 8 is unsupported with .net framework, and yes - using a higher version then 7.3 comes with runtime risks, but there exists polyfill packages to help out.  

Version 4.X will not go any other route, you have the to wait for version 5.  I agree that we could have done a major version breaking change on this, but we are transistioning from a policy where major version only changed very rarely, and included very major changes.  Like version 2 existed for 11 years or so, version 3 existed for about the same time, version 4 will exist for probably 3 years, and after this one we will try to be more semver compatible.  That is the goal.  

So please just regard version 4.6 as having a breaking change wrt version 4.5.  

But, you can try the latest alpha version of version 5 (on myget), that one has the fix with the removal of TestDelegate and ActualValueDelegate.  (link to myget:  https://www.myget.org/feed/nunit/package/nuget/NUnit/5.0.0-alpha.1.19 .   

We will try to get out a beta version soon on nuget.org, but getting some feedback on the alpha would be very helpful.

**Comment by OsirisTerje** [comment]
@manfred-brands @stevenaw What do you think about removing the TestDelegate and ACtualValueDelegate from .net framework build in e.g. version 4.7 ?  We could do that - although I am not sure if it really is so easy.  I removed it from version 5 just now, and it did bite me back on several places.

**Comment by stevenaw** [comment]
@OsirisTerje  I figure all ideas are worth discussing and talking through. Do you recall the issues that came up?

As for removing within 4.x, I think removing it in only the netfx build could complicate projects which multitarget both it and modern .net as their source would then be broken for the non-netfx builds. This might lead us to consider that any removal "might" be simplest if done in all builds.

I don't have an easy answer on the larger question at the moment though I can try to think of other alternatives we can weigh options against.

**Comment by manfred-brands** [comment]
@Otiel  Microsoft is just hedging any complaints. At the same time they have videos that show how to use new language features whilst targeting .NET Framework 
* https://www.youtube.com/watch?v=dLAgn_vljqA
* https://www.youtube.com/watch?v=5eNtdj68qW4

Many language features are "syntactic sugar" that only require a newer compiler to build, not a newer runtime to run.
NUnit source code itself uses LangVersion C#14 even though it targets net462 and net6.0 and net8.0

At my work we multi-target .NET Framework and .NET 8+ and use C#14 features include the new extension syntax and they work for all targets.
The only features that I have found that require runtime support are static members in interfaces and generic attributes.

As an example, nunit has:
```csharp
public TestAdapter Test => _test ??= new(_testExecutionContext.CurrentTest);
```

This uses the C#8.0 `??=` operator and the C#9.0 targeted `new`.
When looking at this code with tools like [ILSpy](https://github.com/icsharpcode/ILSpy/releases/tag/v10.0.1) it shows:
```csharp
public TestAdapter Test => _test ?? (_test = new TestAdapter(_testExecutionContext.CurrentTest));
```
The compiler unwrapped the `??=` operator and filled in the type for the `new`.

The feature that NUnit uses to disambiguate the Action and TestDelegate overloads is the [OverloadResolutionPriority Attribute](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-13.0/overload-resolution-priority)
This allows to compiler to prefer the `Action` overload over the `TestDelegate` even though both match.
It will only take the `TestDelegate` overload if the code has an actual variable of type `TestDelegate`.

> I have some code inspection rules that do not pass anymore with around 13 000 errors

I presume these are suggestions to use modern constructs like pattern matching and targeted new and collections initializers.
Otherwise you could set the `AnalyzerMode` or `AnalyzerLevel` to a lower value.

Also, if you have a separate NUnit project from you code under test, you would only need to set the LangVersion for the test project.

**Comment by hfickes** [comment]
This is breaking our test code which uses .Net Framework 4.8.1
A call to:

  Assert.Throws<System.ArgumentNullException>(() => something(null));

gives a compile time error of:

  The call is ambiguous between the following methods or properties: 'Assert.Throws<TExpected>(TestDelegate)' and 'Assert.Throws<TExpected>(Action)'

Maybe this works in later .Net versions but we are stuck at 4.8 because this is a Microsoft Office VSTO project and Office doesn't permit addins past 4.8.

We can workaround this by casting to an Action such as:

  Assert.Throws<System.ArgumentNullException>((System.Action)(() => something(null)));

but that is inelegant, and I'd have to do it almost 400 times <grin>.
My preferred solution would be to ditch TestDelegate completely.

-Herb
